### DISCUSSION TITLE
Updating pre-secondary structure molecular node blender file to properly utilize secondary structure cartoon

### DISCUSSION THREAD

**Discussion by heejongkim** [BradyAJohnston/MolecularNodes#231] [Q&A]
Hi,

I'm fascinated with the secondary structure feature recently added in molecular node.
I'm just wondering if there's a way to update the existing imported pdb data in the blender somehow so that cartoon node can properly render the  secondary structures retrieved from PDB files. (I would like to use the definitions from PDB rather than computational prediction) I can see that pre-secondary structure molecular node import lacks several columns that exist in the latest version's import.

Since I already made animations with keyframe association, it would be awesome if I can just swap/update only structures to show secondary structures. (and subsequently adjust nodes as well while maintaining keyframes)

Thank you.

best,
heejong

**Comment by BradyAJohnston** [comment]
If both of the structures have the same number of atoms, then it should be relatively straight forward. You can use the `Sample Index` node to get the secondary structure information from a newly imported PDB, and store it in the node tree of the 'old' but already keyframes structure like so:

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/2fd4e3b8-ede7-4153-a52d-727c01e5c182)

If both structures are just imported vied the PDB download button then this should work perfectly :)

**Reply by heejongkim** [reply]
Hi Brady,

After having a couple of days banging my head against a wall, I'm here again.
So, I noticed that Named Attribute node with sec_struct can be retrieved from NEW pdb import. I went ahead and created the node connections as shown above but I'm still wondering how I can "link" or "copy" that secondary structure info to the old imported object. What am I missing here?

Thank you.

best,
heejong

**Reply by BradyAJohnston** [reply]
So what needs to happen is that we need to sample the `NEW_PROTEIN` inside the node tree of the `OLD_PROTEIN` and to create the new attribute `sec_struct` inside of that node tree.


### Newly Created Protein

This is the node tree of the new protein. This protein does have `sec_struct` attribute, but doesn't have any of the animations applied. The output of the node tree is just the mesh. You also don't even need a node tree for this, so you could also just remove the node tree on the new protein.

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/1a78c605-3924-4d65-b0e6-9b52e26f8013)

### Old Protein

The old protein that already has the animations applied, inside of this node tree we need to create and store the new attribute `sec_struct` and we do that with the `Store Named Attribute` node. 

The way we get the correct secondary structure, is by sampling the mesh of the newly created protein. We sample based on index, so as long as each protein hasn't had any modifications to the number of atoms from selections or deletions, each atom in the old protein and the new protein will match based on `Index`. 

Using that index we sample the `sec_struct` attribute from the new protein, and store it as an attribute of the same name on the old protein, which now lets us successfully use the cartoon style node.

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/64259d40-0d19-4d55-a8c8-4ee71b3b690f)

**Reply by Dan-Burns** [reply]
I am not seeing "sec_struct" as an available option in "Named Attribute".  I typed it in but this approach does not seem to work for me.  

Thank you,

Dan

Python 3.10.6
MDAnalysis 2.5.0
Blender 3.6.3

**Reply by BradyAJohnston** [reply]
Hi @Dan-Burns, what import method are you using? Currently `sec_struct` isn't added if importing MD trajectories. 

With the imported structure, if  you don't apply any styles and just look at the underlying data, is there `sec_struct` in the spreadsheet? You can access the spreadsheet in the `Geoemetry Nodes` workspace.

![CleanShot 2023-09-22 at 14 24 10](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/6e80077d-f85f-4549-94a2-cabf0163345a)

**Reply by Dan-Burns** [reply]
Thanks @BradyAJohnston.  When I import the structure after the trajectory import, I can see in the spreadsheet that there is a sec_struct column.  Then I found it in the named attribute dropdown as well but still can't seem to make it work for the trajectory.  If I add the determine secondary structure node, I still cannot get the cartoon representation for the trajectory.  No big deal for now, I may be missing something.
